waku/waku_sync/raw_bindings.nim

```diff
@@ -119,21 +119,67 @@ proc raw_reconcile(
-
 # https://github.com/waku-org/negentropy/blob/d4845b95b5a2d9bee28555833e7502db71bf319f/cpp/negentropy_wrapper.h#L51
 proc raw_reconcile_with_ids(
   negentropy: Negentropy, query: ptr Buffer, r: ptr BindingResult
 ): int {.header: NEGENTROPY_HEADER, importc: "reconcile_with_ids_no_cbk".}
 
+# https://github.com/waku-org/negentropy/blob/d4845b95b5a2d9bee28555833e7502db71bf319f/cpp/negentropy_wrapper.h#L35
+proc free(
+  negentropy: Negentropy
+) {.header: NEGENTROPY_HEADER, importc: "negentropy_delete".}
+
 # https://github.com/waku-org/negentropy/blob/d4845b95b5a2d9bee28555833e7502db71bf319f/cpp/negentropy_wrapper.h#L53
 proc free_result(
   r: ptr BindingResult
 ) {.header: NEGENTROPY_HEADER, importc: "free_result".}
 
+### SubRange ###
+
+type SubRange* = distinct pointer
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L57
+proc subrange_init(
+  storage: Storage, startTimestamp: uint64, endTimestamp: uint64
+): SubRange {.header: NEGENTROPY_HEADER, importc: "subrange_new".}
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L59
+proc free(subrange: SubRange) {.header: NEGENTROPY_HEADER, importc: "subrange_delete".}
+
+# https://github.com/waku-org/negentropy/blob/d4845b95b5a2d9bee28555833e7502db71bf319f/cpp/negentropy_wrapper.h#L31
+proc size*(
+  subrange: SubRange
+): cint {.header: NEGENTROPY_HEADER, importc: "subrange_size".}
+
+### Negentropy with SubRange ###
+
+type NegentropySubRange* = distinct pointer
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L61
+proc constructNegentropyWithSubRange(
+  subrange: SubRange, frameSizeLimit: uint64
+): NegentropySubRange {.header: NEGENTROPY_HEADER, importc: "negentropy_subrange_new".}
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L65
+proc raw_initiate_subrange(
+  negentropy: NegentropySubRange, r: ptr BindingResult
+): int {.header: NEGENTROPY_HEADER, importc: "negentropy_subrange_initiate".}
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L67
+proc raw_reconcile_subrange(
+  negentropy: NegentropySubRange, query: ptr Buffer, r: ptr BindingResult
+): int {.header: NEGENTROPY_HEADER, importc: "reconcile_subrange".}
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L69
+proc raw_reconcile_with_ids_subrange(
+  negentropy: NegentropySubRange, query: ptr Buffer, r: ptr BindingResult
+): int {.header: NEGENTROPY_HEADER, importc: "reconcile_with_ids_subrange_no_cbk".}
+
+# https://github.com/waku-org/negentropy/blob/3044a30e4ba2e218aee6dee2ef5b4a4b6f144865/cpp/negentropy_wrapper.h#L63
+proc free(
+  negentropy: NegentropySubRange
+) {.header: NEGENTROPY_HEADER, importc: "negentropy_subrange_delete".}
+
```

Comment:
We can't avoid this duplication because of C++ templating ? If not we may be able to "hide" it with some kind of interface.

Reply:
Yup, since wrapper is of C..we need to have separate methods for each type for template code.